app/helpers/random_assign_helper.rb

```diff
@@ -0,0 +1,174 @@
+
+  private
+  # Make sure there's enough reviewee groups in the following list such that
+  # we have enough repeats of 'reviewee_groups' that we can assign every
+  # reviewer group at least 'num_groups_for_reviewers' groups.
+  def generate_shuffled_reviewees(reviewer_groups, reviewee_groups)
+    num_times_to_add_reviewee = (reviewer_groups.size.to_f * @num_groups_for_reviewers / reviewer_groups.size).ceil
+    num_times_to_add_reviewee.times { @shuffled_reviewees += reviewee_groups }
+    @shuffled_reviewees.shuffle
+
+    @shuffled_reviewees_pr = @shuffled_reviewees.map { nil }
+  end
+
+  # We need to get all the existing peer reviews for assignments so we can
+  # skip some assignments if they exist.
+  def remove_groups_from_shuffle_having_peer_review(pr_assignment)
+    pr_assignment.pr_peer_reviews.each do |peer_review|
+      reviewee = peer_review.reviewee
+      @reviewers_assigned_to[peer_review.reviewer.id].add(reviewee.id)
+      remove_reviewee_once_from_shuffled_list(reviewee)
+    end
+  end
+
+  # Required so that existing reviews don't cause a lopsided distribution.
+  def remove_reviewee_once_from_shuffled_list(reviewee)
+    index = @shuffled_reviewees.find_index(reviewee)
+
+    # It is possible that multiple peer reviews exist for the reviewe, and then
+    # are not in the list. Since we've removed all occurences, ignoring is okay.
+    unless index.nil?
+      @shuffled_reviewees.delete_at(index)
+      @shuffled_reviewees_pr.pop()
+    end
+  end
+
+  def perform_assignments
+    shuffle_index = 0
+    remove_ineligible_reviewers()
+
+    while @eligible_reviewers.any?
+      @eligible_reviewers.each do |reviewer|
+        reviewee = get_next_reviewee_from_forward_search(reviewer, shuffle_index)
+
+        # A nil reviewee means we could not find one going forwards, and if so,
+        # then try going backwards through assignments (failure will throw).
+        if not reviewee.nil?
+          add_peer_review_to_db_and_remember_assignment(reviewer, reviewee, shuffle_index)
+        else
+          find_and_swap_with_group_or_throw(reviewer, shuffle_index)
+        end
+
+        shuffle_index += 1
+      end
+
+      remove_ineligible_reviewers()
+    end
+  end
+
+  def remove_ineligible_reviewers
+    @eligible_reviewers.delete_if do |reviewer|
+      @reviewers_assigned_to[reviewer.id].size >= @num_groups_for_reviewers
+    end
+  end
+
+  # Note: This returns nil if forward searching failed (signal to the caller).
+  def get_next_reviewee_from_forward_search(reviewer, shuffle_index)
+    # First, check if the current index provided works.
+    potential_reviewee = @shuffled_reviewees[shuffle_index]
+    if eligible_to_be_assigned(reviewer, potential_reviewee)
+      return potential_reviewee
+    end
+
+    # Since the index doesnt work, look at the next element until the end of
+    # the list, and swap if we find a working group.
+    ((shuffle_index + 1)...@shuffled_reviewees.size).each do |new_shuffle_index|
```

Comment:
Alrighty. Let's change this loop so that it effectively merges in `find_and_swap_with_group_or_throw` too:

```ruby
@shuffled_reviewees.size.times do |i|
  new_shuffle_index = shuffle_index + i % @shuffled_reviewees.size
  ...
  if eligible_to_be_assigned(reviewer, potential_reviewee)
    if new_shuffle_index == shuffle_index  # keep same position
      ...
    elsif new_shuffle_index > shuffle_index  # swap with empty reviewee
      ...
    else  # logic from find_and_swap_with_group_or_throw
      ...
```

If you find this makes the method too long, you can use helpers where appropriate. But only this method should be looping over `@shuffled_reviewers`